File: ios/StockbookTests/LastPaidTests.swift
import Testing
import Foundation
@testable import Stockbook

/// How long an account has gone without money moving.
///
/// The twin of `LastPaidTests.kt`, test for test. Home has always said who owes
/// and how much, and never how long. What this pins is that the clock is started
/// and stopped by **money only** — the two things that reduce a balance without
/// anybody paying, a credit note and a balance transfer, must leave it running.
/// Getting that wrong tells the owner they were paid by somebody who has not paid
/// them since spring, and they stop chasing it.
@MainActor
@Suite("Last paid")
struct LastPaidTests {

    private func makeStore() -> StockbookStore {
        StockbookStore(repository: InMemoryRepository())
    }

    /// 2026-08-27T09:00:00Z, fixed so the arithmetic below is arithmetic and not
    /// a function of when the suite happens to run.
    private let now = Date(timeIntervalSince1970: 1_787_821_200)

    private func daysAgo(_ days: Int) -> Date {
        now.addingTimeInterval(-Double(days) * 86_400)
    }

    private func shopWithProduct() -> (StockbookStore, Product) {
        let store = makeStore()
        let lock = store.addProduct(name: "Cisa lock", stock: 100, cost: 60, price: 95)
        return (store, lock)
    }

    // MARK: The rule on its own

    @Test("With no date to count from it says nothing rather than guessing")
    func noDate() {
        #expect(LastPaid.daysSince(lastPaidAt: nil, since: nil, now: now) == nil)
    }

    @Test("Never paid counts from the day the trading started")
    func fromFirstBill() {
        #expect(LastPaid.daysSince(lastPaidAt: nil, since: daysAgo(40), now: now) == 40)
    }

    @Test("A payment overrides the start date")
    func paymentWins() {
        #expect(
            LastPaid.daysSince(lastPaidAt: daysAgo(5), since: daysAgo(400), now: now) == 5,
            "the clock restarts when money comes in, whatever came before"
        )
    }

    @Test("A clock that has run backwards floors at zero")
    func backwards() {
        // A phone whose date was wrong when a bill was written. Nothing useful to
        // say, but "-3 days ago" on a counter screen is worse than "today".
        let future = now.addingTimeInterval(3 * 86_400)
        #expect(LastPaid.daysSince(lastPaidAt: future, since: nil, now: now) == 0)
    }

    @Test("Nothing is said until it is worth saying")
    func threshold() {
        #expect(!LastPaid.worthSaying(nil))
        #expect(!LastPaid.worthSaying(0))
        #expect(!LastPaid.worthSaying(LastPaid.worthSayingAfterDays - 1))
        #expect(LastPaid.worthSaying(LastPaid.worthSayingAfterDays))
    }

    // MARK: What the store works out from a real book

    @Test("Paying at the counter counts as being paid")
    func counterPayment() throws {
        // A shop whose customers settle on the spot writes no payment rows at
        // all. Reading only those would call every one of them a non-payer.
        let (store, lock) = shopWithProduct()
        store.saveBill(
            lines: [.init(productUID: lock.uid, qty: 1, price: 95)],
            customer: "Ahmed",
            paid: 95,
            createdAt: daysAgo(3)
        )

        let ahmed = try #require(store.customer(key: "ahmed"))
        #expect(ahmed.hasEverPaid)
        #expect(ahmed.quietDays(now: now) == 3)
    }

    @Test("A bill nobody paid anything on does not restart the clock")
    func unpaidBill() throws {
        let (store, lock) = shopWithProduct()
        store.saveBill(
            lines: [.init(productUID: lock.uid, qty: 1, price: 95)],
            customer: "Ahmed",
            paid: 0,
            createdAt: daysAgo(60)
        )

        let ahmed = try #require(store.customer(key: "ahmed"))
        #expect(!ahmed.hasEverPaid, "nothing has come in from them yet")
        #expect(ahmed.quietDays(now: now) == 60, "counted from the bill, which is all there is")
    }

    @Test("The most recent money is the one that counts")
    func mostRecent() throws {
        let (store, lock) = shopWithProduct()
        store.saveBill(
            lines: [.init(productUID: lock.uid, qty: 5, price: 95)],
            customer: "Ahmed",
            paid: 0,
            createdAt: daysAgo(90)
        )
        store.recordPayment(customerKey: "ahmed", amount: 100, receivedAt: daysAgo(70))
        store.recordPayment(customerKey: "ahmed", amount: 100, receivedAt: daysAgo(45))

        #expect(try #require(store.customer(key: "ahmed")).quietDays(now: now) == 45)
    }

    /// The one that would have been wrong in the worst way.
    ///
    /// A credit note reduces what somebody owes and no money changes hands. If it
    /// reset this clock, a customer written off in part would read as having just
    /// paid — and the owner would stop chasing the rest.
    @Test("A credit note is not a payment")
    func creditNoteIsNotPayment() throws {
        let (store, lock) = shopWithProduct()
        store.saveBill(
            lines: [.init(productUID: lock.uid, qty: 5, price: 95)],
            customer: "Ahmed",
            paid: 0,
            createdAt: daysAgo(80)
        )
        store.addCreditNote(customerKey: "ahmed", amount: 200, issuedAt: daysAgo(2))

        let ahmed = try #require(store.customer(key: "ahmed"))
        #expect(!ahmed.hasEverPaid)
        #expect(ahmed.quietDays(now: now) == 80, "still eighty days without a coin")
    }

    /// The same rule, for the other thing that moves a balance without money.
    @Test("A balance transfer is not a payment")
    func transferIsNotPayment() throws {
        let (store, lock) = shopWithProduct()
        store.addCustomer(name: "Ahmed Riyadh")
        store.saveBill(
            lines: [.init(productUID: lock.uid, qty: 5, price: 95)],
            customer: "Ahmed Jeddah",
            paid: 0,
            createdAt: daysAgo(80)
        )
        _ = store.transferBalance(
            fromKey: "ahmed jeddah",
            intoKey: "ahmed riyadh",
            amount: 200,
            movedAt: daysAgo(2)
        )

        let jeddah = try #require(store.customer(key: "ahmed jeddah"))
        #expect(!jeddah.hasEverPaid)
        #expect(jeddah.quietDays(now: now) == 80)

        // And the end it arrived at was not paid either — it was charged.
        let riyadh = try #require(store.customer(key: "ahmed riyadh"))
        #expect(!riyadh.hasEverPaid)
    }

    @Test("An opening balance alone has no date and says nothing")
    func openingBalanceOnly() throws {
        let store = makeStore()
        store.addCustomer(name: "Ahmed", openingBalance: 500)

        let ahmed = try #require(store.customer(key: "ahmed"))
        #expect(
            ahmed.quietDays(now: now) == nil,
            "carried over from the paper book with no date to count from"
        )
    }

    // MARK: The other side of the book

    @Test("A supplier is how long since the shop paid them")
    func supplierSide() throws {
        let (store, product) = shopWithProduct()
        let supplier = try #require(store.addSupplier(name: "Al Faisal"))
        store.recordPurchase(
            product: product,
            supplierKey: supplier.key,
            quantity: 10,
            unitCost: 40,
            paid: 0,
            createdAt: daysAgo(50)
        )

        let unpaid = try #require(store.supplier(key: supplier.key))
        #expect(!unpaid.hasEverPaid)
        #expect(unpaid.quietDays(now: now) == 50)

        store.recordSupplierPayment(supplierKey: supplier.key, amount: 200, paidAt: daysAgo(10))

        let paid = try #require(store.supplier(key: supplier.key))
        #expect(paid.hasEverPaid)
        #expect(paid.quietDays(now: now) == 10)
    }
}

File: android/core/src/test/kotlin/com/stockbook/core/LastPaidTests.kt
package com.stockbook.core

import com.stockbook.core.model.LastPaid
import com.stockbook.core.store.DraftLine
import com.stockbook.core.store.InMemoryRepository
import com.stockbook.core.store.StockbookStore
import java.time.Instant
import java.time.temporal.ChronoUnit
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * How long an account has gone without money moving.
 *
 * Home has always said who owes and how much, and never how long. What this
 * pins is that the clock is started and stopped by **money only** — the two
 * things that reduce a balance without anybody paying, a credit note and a
 * balance transfer, must leave it running. Getting that wrong tells the owner
 * they were paid by somebody who has not paid them since spring, and they stop
 * chasing it.
 */
class LastPaidTests {

    private fun store() = StockbookStore(InMemoryRepository())

    private val now: Instant = Instant.parse("2026-08-27T09:00:00Z")
    private fun daysAgo(days: Long): Instant = now.minus(days, ChronoUnit.DAYS)

    private fun shopWithProduct(): Pair<StockbookStore, String> {
        val store = store()
        val lock = store.addProduct("Cisa lock", 100, 60.0, 95.0)
        return store to lock.uid
    }

    // --- The rule on its own

    @Test
    fun `with no date to count from it says nothing rather than guessing`() {
        assertNull(LastPaid.daysSince(lastPaidAt = null, since = null, now = now))
    }

    @Test
    fun `never paid counts from the day the trading started`() {
        assertEquals(40, LastPaid.daysSince(lastPaidAt = null, since = daysAgo(40), now = now))
    }

    @Test
    fun `a payment overrides the start date`() {
        assertEquals(
            5,
            LastPaid.daysSince(lastPaidAt = daysAgo(5), since = daysAgo(400), now = now),
            "the clock restarts when money comes in, whatever came before"
        )
    }

    @Test
    fun `a clock that has run backwards floors at zero`() {
        // A phone whose date was wrong when a bill was written. Nothing useful
        // to say, but "-3 days ago" on a counter screen is worse than "today".
        assertEquals(0, LastPaid.daysSince(lastPaidAt = now.plus(3, ChronoUnit.DAYS), since = null, now = now))
    }

    @Test
    fun `nothing is said until it is worth saying`() {
        assertFalse(LastPaid.worthSaying(null))
        assertFalse(LastPaid.worthSaying(0))
        assertFalse(LastPaid.worthSaying(LastPaid.WORTH_SAYING_AFTER_DAYS - 1))
        assertTrue(LastPaid.worthSaying(LastPaid.WORTH_SAYING_AFTER_DAYS))
    }

    // --- What the store works out from a real book

    @Test
    fun `paying at the counter counts as being paid`() {
        // A shop whose customers settle on the spot writes no payment rows at
        // all. Reading only those would call every one of them a non-payer.
        val (store, lock) = shopWithProduct()
        store.saveBill(listOf(DraftLine(lock, 1, 95.0)), "Ahmed", paid = 95.0, createdAt = daysAgo(3))

        val ahmed = assertNotNull(store.customer("ahmed"))
        assertTrue(ahmed.hasEverPaid)
        assertEquals(3, ahmed.quietDays(now))
    }

    @Test
    fun `a bill nobody paid anything on does not restart the clock`() {
        val (store, lock) = shopWithProduct()
        store.saveBill(listOf(DraftLine(lock, 1, 95.0)), "Ahmed", paid = 0.0, createdAt = daysAgo(60))

        val ahmed = assertNotNull(store.customer("ahmed"))
        assertFalse(ahmed.hasEverPaid, "nothing has come in from them yet")
        assertEquals(60, ahmed.quietDays(now), "counted from the bill, which is all there is")
    }

    @Test
    fun `the most recent money is the one that counts`() {
        val (store, lock) = shopWithProduct()
        store.saveBill(listOf(DraftLine(lock, 5, 95.0)), "Ahmed", paid = 0.0, createdAt = daysAgo(90))
        store.recordPayment("ahmed", 100.0, receivedAt = daysAgo(70))
        store.recordPayment("ahmed", 100.0, receivedAt = daysAgo(45))

        assertEquals(45, assertNotNull(store.customer("ahmed")).quietDays(now))
    }

    /**
     * The one that would have been wrong in the worst way.
     *
     * A credit note reduces what somebody owes and no money changes hands. If it
     * reset this clock, a customer written off in part would read as having just
     * paid — and the owner would stop chasing the rest.
     */
    @Test
    fun `a credit note is not a payment`() {
        val (store, lock) = shopWithProduct()
        store.saveBill(listOf(DraftLine(lock, 5, 95.0)), "Ahmed", paid = 0.0, createdAt = daysAgo(80))
        store.addCreditNote("ahmed", amount = 200.0, issuedAt = daysAgo(2))

        val ahmed = assertNotNull(store.customer("ahmed"))
        assertFalse(ahmed.hasEverPaid)
        assertEquals(80, ahmed.quietDays(now), "still eighty days without a coin")
    }

    /** The same rule, for the other thing that moves a balance without money. */
    @Test
    fun `a balance transfer is not a payment`() {
        val (store, lock) = shopWithProduct()
        store.addCustomer("Ahmed Riyadh")
        store.saveBill(listOf(DraftLine(lock, 5, 95.0)), "Ahmed Jeddah", paid = 0.0, createdAt = daysAgo(80))
        store.transferBalance("ahmed jeddah", "ahmed riyadh", 200.0, movedAt = daysAgo(2))

        val jeddah = assertNotNull(store.customer("ahmed jeddah"))
        assertFalse(jeddah.hasEverPaid)
        assertEquals(80, jeddah.quietDays(now))

        // And the end it arrived at was not paid either — it was charged.
        val riyadh = assertNotNull(store.customer("ahmed riyadh"))
        assertFalse(riyadh.hasEverPaid)
    }

    @Test
    fun `an opening balance alone has no date and says nothing`() {
        val store = store()
        store.addCustomer("Ahmed", openingBalance = 500.0)

        val ahmed = assertNotNull(store.customer("ahmed"))
        assertNull(ahmed.quietDays(now), "carried over from the paper book with no date to count from")
    }

    // --- The other side of the book

    @Test
    fun `a supplier is how long since the shop paid them`() {
        val (store, _) = shopWithProduct()
        val product = assertNotNull(store.products.firstOrNull())
        val supplier = assertNotNull(store.addSupplier("Al Faisal"))
        store.recordPurchase(product, supplier.key, quantity = 10, unitCost = 40.0, paid = 0.0, createdAt = daysAgo(50))

        val unpaid = assertNotNull(store.supplier(supplier.key))
        assertFalse(unpaid.hasEverPaid)
        assertEquals(50, unpaid.quietDays(now))

        store.recordSupplierPayment(supplier.key, 200.0, paidAt = daysAgo(10))

        val paid = assertNotNull(store.supplier(supplier.key))
        assertTrue(paid.hasEverPaid)
        assertEquals(10, paid.quietDays(now))
    }
}
